docs/ksshell.1
.TH "KSSHELL" "1"
.SH "NAME"
ksshell \(em an interactive kickstart shell
.SH "SYNOPSIS"
.PP
\fBksflatten\fR [\fB\-i\fR | \fB\-\-input INFILE\fP]  [\fB\-o\fR | \fB\-\-output OUTFILE\fP]  [\fB\-v\fR | \fB\-\-version VERSION\fP]
.SH "DESCRIPTION"
.PP
\fBksshell\fR is an interactive kickstart shell.  It optionally takes an input kickstart file as the basis,
allows the user to specify additional kickstart commands, and then writes out the finished kickstart file
to stdout or the given file name.  This program supports all the usual readline niceties including tab
completion of kickstart commands and their options, though not the values those options can take.
.PP
In addition to understanding all the kickstart commands, \fBksshell\fR has some builtin commands of its
own to make working with kickstart files in the context of a shell easier:
.IP .clear
Clear the existing kickstart data, including any from INFILE.  This essentially starts you over from a
blank state.
.IP .quit
Quit the interactive shell, either saving to the file given by OUTFILE or printing to stdout if none
was given.
.IP .show
Print the current kickstart file state.
.SH "EXIT STATUS"
.PP
\fBksflatten\fR returns 0 on success, and 1 if VERSION is incorrect.  If INFILE does not exist, a warning
will be printed but the user will still be dumped to the interactive shell.
.SH "OPTIONS"
.IP "\fB\-i\fR, \fB\-\-input INFILE\fP" 10
The name of the input kickstart file.
.IP "\fB\-o\fR, \fB\-\-output OUTFILE\fP" 10
Write the flattened kickstart file to OUTFILE, or stdout if no filename is given.
.IP "\fB\-v\fR, \fB\-\-version VERSION\fP" 10
Use this version of kickstart syntax when processing the file, or the latest if no version is given.
.SH "SEE ALSO"
.PP
ksflatten (1), ksvalidator (1), ksverdiff (1)


ChangeLog
2008-04-16  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* Makefile:  Update for git.

	* pykickstart/parser.py (KickstartHandlers.doNetwork): query is now a valid
	bootproto for networking (#441837).

2006-12-15  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version (only translations have
	changed since the last rebuild - #216620).

2006-12-05  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-12-04  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doRepo):  Fix traceback
	(#218274).

2006-12-01  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/parser.py (KickstartHandlers.deprecatedCommands):  Fix
	traceback (#218047, #218059).

2006-11-30  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-11-21  Chris Lumens  <clumens@redhat.com>

	* po/pykickstart.pot:  Updated (#216620).

	* pykickstart/parser.py:  Update string formats to make translation
	easier.

2006-11-14  Chris Lumens  <clumens@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doIscsi):  Use required
	attribute.

2006-11-09  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doLogging):  Add --level.

2006-10-24  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* Makefile:  Fix for new Release format.

2006-10-17  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py:  Add writer for key command.

2006-10-17  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version.
	
	* pykickstart/parser.py, pykickstart/constants.py: add --skip for key

2006-10-16  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version.

	* pykickstart/parser.py (KickstartHandlers.doKey): Support for
	installation key
	* pykickstart/data.py (KickstartData.__init__): Likewise.

2006-10-13  Bill Nottingham  <notting@redhat.com>

	* af.po, am.po, ar.po, be.po, bg.po, bn.po, cy.po, et.po,
	  eu_ES.po, fa.po, fi.po, he.po, hy.po, kn.po, ku.po, lo.po, lt.po,
	  lv.po, mk.po, my.po, nn.po, no.po, nso.po, ro.po, si.po, sq.po, th.po,
	  ur.po, vi.po, zu.po: Use a valid charset (#210720)
	
	* pykickstart.spec, setup.py: Bump version


2006-09-29  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doMethod):  Fix traceback
	and test (#208557).

2006-09-25  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doMethod):  Add support for
	--biospart option (#207585).
	* pykickstart/writer.py (KickstartWriter.doMethod):  Likewise.

	* pykickstart/writer.py (KickstartWriter.doIscsi):  Output new syntax.
	(KickstartWriter.doZFCP):  Likewise.
	(KickstartWriter.doIscsiName):  Added.

2006-09-20  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doIscsi): New and
	improved iscsi syntax.  Also, allow multiple zfcp devs
	* pykickstart/data.py: Likewise.

2006-07-13  Chris Lumens  <clumens@redhat.com>

	* pykickstart/data.py:  Add noipv4 and noipv6 to KickstartNetworkData.
	* pykickstart/parser.py (KickstartHandlers.doNetwork):  Limit
	bootproto choices to what anaconda can handle.  Add support for
	--noipv4 and --noipv6 options.
	* pykickstart/writer.py (KickstartWriter.doNetwork):  Write out new
	networking options.

2006-06-20  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/writer.py (KickstartWriter.doIscsi):  Add iscsi writer.

2006-06-20  Jeremy Katz  <katzj@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doIscsi): Initial iscsi
	support.  This syntax is likely to change but this makes testing
	so much easier.
	* pykickstart/data.py (KickstartData.__init__): Add iscsi data.

2006-06-20  Chris Lumens  <clumens@redhat.com>

	* pykickstart/parser (KickstartHandlers.doRaid):  Handle a RAID device
	as an integer instead of a string (#176537).

2006-06-16  Jeremy Katz  <katzj@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doMethod): Handle nfs --opts

2006-06-06  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec:  Fix BuildRequires (#194156).

2006-05-25  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-05-24  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter):  Change order of LVM-related
	functions (#193073).

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec:  Require urlgrabber now.

	* po/*:  Update translation files.

	* pykickstart/parser.py (KickstartParser.handleCommand):  Return a
	more useful error message on unknown command.

	* pykickstart/writer.py (KickstartWriter.doLogicalVolume):  Fix typo.

	* validator/ksvalidator:  Accept a URL in addition to a file name.

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doPackages):  Don't write out
	an empty packages section, since we may want this step to be
	interactive (#192851).

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/writer.py (KickstartWriter.doFirewall):  Rename --ports
	to match the parser.

2006-05-17  Peter Jones  <pjones@redhat.com>

	* pykickstart/parser.py: Add "multipath" command and doMultiPath
	handler.
	* pykickstart/data.py: Add kickstartMpPathData and
	KickstartMultiPathData objects to handle multipath info, as well as
	KickstartData.mpaths .
	* pykickstart/writer.py: add doMultiPath
	* docs/programmers-guide: add completely useless documentation for
	multipath kickstart objects.

2006-05-15  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/data.py (KickstartNetworkData):  Add MTU (#191328).

tools/ksshell.py
#!/usr/bin/env python3
#
# Chris Lumens <clumens@redhat.com>
#
# Copyright 2013-2014 Red Hat, Inc.
#
# This copyrighted material is made available to anyone wishing to use, modify,
# copy, or redistribute it subject to the terms and conditions of the GNU
# General Public License v.2.  This program is distributed in the hope that it
# will be useful, but WITHOUT ANY WARRANTY expressed or implied, including the
# implied warranties of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  Any Red Hat
# trademarks that are incorporated in the source code or documentation are not
# subject to the GNU General Public License and may only be used or replicated
# with the express permission of Red Hat, Inc.

# This script takes as input an optional input kickstart file and an optional
# kickstart syntax version (the latest is assumed, if none specified).  It
# then provides an interactive shell for the user to manipulate the state of
# the input kickstart file and either prints the final results on stdout or to
# a designated output file when the program quits.

# TODO:
# - error reporting always says you are on line 1
# - handle sections like %packages
# - some meta-commands might be nice:
#   - .delete (a previous line)
#   - .help (requires moving help text into each optionparser object?)
#   - .save

# pylint: disable=found-_-in-module-class

import readline
import argparse
import os, six, sys

from pykickstart.i18n import _
from pykickstart.errors import KickstartError, KickstartVersionError
from pykickstart.parser import KickstartParser, preprocessKickstart
from pykickstart.version import DEVEL, makeVersion

##
## INTERNAL COMMANDS
## These are commands that control operation of the kickstart shell and are
## handled by this program.  They are not recognized kickstart commands.
##

class InternalCommand(object):
    def __init__(self):
        self.op = argparse.ArgumentParser()

    def execute(self, parser):
        pass

class ClearCommand(InternalCommand):
    def execute(self, parser):
        version = parser.version
        parser.handler = makeVersion(version)

class QuitCommand(InternalCommand):
    def execute(self, parser):
        raise EOFError

class ShowCommand(InternalCommand):
    def execute(self, parser):
        print(parser.handler)

##
## COMMAND COMPLETION
##

class KickstartCompleter(object):
    def __init__(self, handler, internal):
        # Build up a dict of kickstart commands and their valid options:
        # { command_name: [options] }
        self.commands = {}

        for (cStr, cObj) in handler.commands.items():
            self._add_command(cStr, cObj)

        for (cStr, cObj) in internal.items():
            self._add_command(cStr, cObj)

        self.currentCandidates = []

    def _add_command(self, cStr, cObj):
        self.commands[cStr] = []

        # Internal and simple commands don't have any optparse crud.
        if cStr.startswith(".") or not hasattr(cObj, "op"):
            return

        for opt in cObj.op.option_list:
            self.commands[cStr] += opt._short_opts + opt._long_opts

    def complete(self, _text, state):
        response = None

        # This is the first time Tab has been pressed, so build up a list of matches.
        if state == 0:
            origline = readline.get_line_buffer()
            begin = readline.get_begidx()
            end = readline.get_endidx()

            beingCompleted = origline[begin:end]
            parts = origline.split()

            if not parts:
                # Everything's a match for an empty string.
                self.currentCandidates = sorted(self.commands.keys())
            else:
                try:
                    # Ignore leading whitespace when trying to figure out
                    # completions for a kickstart command.
                    if begin == 0 or origline[:begin].strip() == "":
                        # first word
                        candidates = self.commands.keys()
                    else:
                        # later word
                        candidates = self.commands[parts[0]]

                    if beingCompleted:
                        self.currentCandidates = [w for w in candidates if w.startswith(beingCompleted)]
                    else:
                        self.currentCandidates = list(candidates)
                except (KeyError, IndexError):
                    self.currentCandidates = []

        try:
            response = self.currentCandidates[state]
        except IndexError:
            response = None

        return response

def main():

    ##
    ## OPTION PROCESSING
    ##

    op = argparse.ArgumentParser()
    op.add_argument("-i", "--input", dest="input",
                    help=_("a basis file to use for seeding the kickstart data (optional)"))
    op.add_argument("-o", "--output", dest="output",
                    help=_("the location to write the finished kickstart file, or stdout if not given"))
    op.add_argument("-v", "--version", dest="version", default=DEVEL,
                    help=_("version of kickstart syntax to validate against"))

    opts = op.parse_args(sys.argv[1:])

    ##
    ## SETTING UP PYKICKSTART
    ##

    try:
        kshandler = makeVersion(opts.version)
    except KickstartVersionError:
        print(_("The version %s is not supported by pykickstart") % opts.version)
        sys.exit(1)

    ksparser = KickstartParser(kshandler, followIncludes=True, errorsAreFatal=False)

    if opts.input:
        try:
            processedFile = preprocessKickstart(opts.input)
            ksparser.readKickstart(processedFile)
            os.remove(processedFile)
        except KickstartError as e:
            # Errors should just dump you to the prompt anyway.
            print(_("Warning:  The following error occurred when processing the input file:\n%s\n") % e)

    internalCommands = {".clear": ClearCommand(),
                        ".show": ShowCommand(),
                        ".quit": QuitCommand()}

    ##
    ## SETTING UP READLINE
    ##

    readline.parse_and_bind("tab: complete")
    readline.set_completer(KickstartCompleter(kshandler, internalCommands).complete)

    # Since everything in kickstart looks like a command line arg, we need to
    # remove '-' from the delimiter string.
    delims = readline.get_completer_delims()
    readline.set_completer_delims(delims.replace('-', ''))

    ##
    ## REPL
    ##

    print("Press ^D to exit.")

    while True:
        try:
            line = six.moves.input("ks> ")  # pylint: disable=no-member
        except EOFError:
            # ^D was hit, time to quit.
            break
        except KeyboardInterrupt:
            # ^C was hit, time to quit.  Don't be like other programs.
            break

        # All internal commands start with a ., so if that's the beginning of the
        # line, we need to dispatch ourselves.
        if line.startswith("."):
            words = line.split()
            if words[0] in internalCommands:
                try:
                    internalCommands[words[0]].execute(ksparser)
                except EOFError:
                    # ".quit" was typed, time to quit.
                    break
            else:
                print(_("Internal command %s not recognized.") % words[0])

            continue

        # Now process the line of input as if it were a kickstart file - just an
        # extremely short one.
        try:
            ksparser.readKickstartFromString(line)
        except KickstartError as e:
            print(e)

    # And finally, print the output kickstart file.
    if opts.output:
        with open(opts.output, "w") as fd:
            fd.write(str(ksparser.handler))
    else:
        print("\n" + str(ksparser.handler))

if __name__ == "__main__":
    main()
